dvc/repo/__init__.py

```diff
@@ -375,8 +375,8 @@ def stages(self):
         return self.stage.collect_repo(onerror=error_handler)
 
     def find_outs_by_path(self, path, outs=None, recursive=False, strict=True):
-        if not outs:
-            outs = [out for stage in self.stages for out in stage.outs]
+        # using `outs_graph` to ensure graph checks are run
+        outs = outs or self.outs_graph
```

Comment:
The disadvantage of it might be that, for example, `dvc.api.open` might start giving these unwanted errors, if their graphs are not correct.

Maybe, instead of giving these errors at all the times, we should only error out if `n(outs) > 1`?

Reply:
But maybe it's not the tree that needs to worry about this.

Reply:
Also, this change is unnecessary to solve #5027, but for #4010.

Reply:
There are lots of assumptions around the code about our outs/stages being in-check with the proper DAG, so we indeed need to check the graph. There were some discussions around whether or not some of the dag checks are really that necessary (e.g. overlapping outputs might be used to `dvc checkout` particular versions of datasets on demand), but so far there wasn't a good scenario that people were actively asking for.